### ISSUE TITLE
Running sage output in percolator

### ISSUE THREAD

**Issue by timosachsenberg** [lazear/sage#53]
Hi,
thanks for your great tool and for making it open-source!
I just ran into a small issue. If I call percolator on the sage output I get an error:

```
Started Mon Mar 20 10:45:47 2023
Hyperparameters: selectionFdr=0.01, Cpos=0, Cneg=0, maxNiter=10
Reading tab-delimited input from datafile results.sage.tsv

WARNING: Tab delimited input does not contain ScanNr column,
         scan numbers will be assigned automatically.

Features:
num_proteins
Exception caught: ERROR: Reading tab file, error reading PSM on line 2. Could not read label.
```

could it be percolator is case-sensitive and doesn't recognize "scannr"? Probably same for other columns?

**Comment by lazear** [comment]
Hi Timo,

IIRC percolator is case sensitive - I generally used Mokapot instead, which was a bit more flexible in terms of file format. (Percolator also requires "proteins" to be the final column, but that the list of proteins are tab delimited themselves. This obviously breaks compatibility with standard TSV semantics/parsers, which was another point in favor of Mokapot).

Its also possible that compatibility with Mokapot/Percolator has shifted over time 🙃 - let me do some testing. I'm happy to add an additional output format that is Percolator compatible

**Comment by lazear** [comment]
@timosachsenberg  Which version of percolator are you targeting, and what flags are you using? I have started working on this and want to make sure we're compatible. Is a CLI flag for Sage sufficient (e.g. `sage config.json --write-pin`)?

Also, have you considered just skipping percolator and using the default LDA scoring from Sage? Initial testing (on 20-file IonStar dataset) shows that LDA outperforms percolator (and is much faster) - but perhaps I am not using the proper percolator flags

**Comment by timosachsenberg** [comment]
A command line flag is totally sufficient.
I have to admit that I did not experiment a lot with the LDA option and I don’t know about the implementation details. e.g. did you test this on entrapment datasets to make sure it does not overfit? I would be a bit surprised/excited if it performs in general better than the C-SVM. I hope to be able to look into it soon.

**Comment by lazear** [comment]
I wouldn't get too excited - by "outperform" I mean within a percentage point or two (and I'm sure this is dataset or search-parameter dependent, and there are cases where percolator outperforms)

Here are the results from an entrapment search of a single file from [PXD001468](http://proteomecentral.proteomexchange.org/cgi/GetDataset?ID=PXD001468) (20742 human sequences, 100373 non-human sequences [plant, archae, bacteria, viruses])

![image](https://user-images.githubusercontent.com/14057244/230644077-ee21fa5e-bace-4cf3-841e-8ff9281010d1.png)


![image](https://user-images.githubusercontent.com/14057244/230644164-15c1d3f6-a711-4ae0-9940-ed9af35c83ba.png)

**Comment by lazear** [comment]
And here's another entrapment search on the IonStar dataset (PXD003881, 20 files, human + ecoli, ~633k target PSMs q<0.01):
![image](https://user-images.githubusercontent.com/14057244/230654189-bebcae19-91ff-4c69-a4a4-3bb47d30a081.png)

**Comment by fcyu** [comment]
In calculate the false match rate, did you multiple something like `#{human peptides in database} / #{ecoli peptides in database}`? If not, the false discovery rate is underestimated.

Best,

Fengchao

**Comment by lazear** [comment]
Hi Fengchao,

Thanks for the input - I am using the definition from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5374549/

You're correct about the need for correcting for entrapment database size (~5x in this case) for an accurate FMR - not sure if that is the correct formula. I think it's `(cumulative # of entrapments/total hits) * (# of all database sequences/# of entrapment sequences)` (assuming false hits are iid among the database, they have a ~80% chance of being drawn from the entrapment sequences) - but don't hold me to it 😄

 I'm also not sure that it is completely relevant to the question at hand, which is whether LDA is bringing through more entrapment sequences than SVM (given that LDA q-value ~ SVM q-value, it is not)

Here is the above plot corrected (just using # of proteins, rather than tryptic sequences, because this is just a heuristic after all)
![image](https://user-images.githubusercontent.com/14057244/230666740-07a7afea-3947-4579-b69e-85b928708568.png)

**Comment by fcyu** [comment]
> I think it's (cumulative # of entrapments/total hits) * (# of all database sequences/# of entrapment sequences) (assuming false hits are iid among the database, they have a ~80% chance of being drawn from the entrapment sequences) - but don't hold me to it

I also find it hard to figure out the very precise multiplication factor because it relates to 1) how to perform the search (separated or concatnated); 2) how to calculate the FDR ( `d/t` or `2d/(d+t)`); and 3) if excluding the entrapment hits in the calculation... I would rather not jump into that rabbit hole now, so I said "multiple **something like**..." ;)


> I'm also not sure that it is completely relevant to the question at hand, which is whether LDA is bringing through more entrapment sequences than SVM (given that LDA q-value ~ SVM q-value, it is not)

Apologies for the diversion. I couldn't help but think of this question when I saw your appearance.

Best,

Fengchao

**Comment by lazear** [comment]
Thanks for the catch, definitely an oversight on my part!

**Comment by lazear** [comment]
@timosachsenberg, the new release build should have percolator support, please let me know if it works!

**Comment by timosachsenberg** [comment]
Thanks! Output looks good at first glance... I will probably next week find some time to complete the parser.

**Comment by timosachsenberg** [comment]
hmm Charge columns probably need to be 1-hot encoded. Did you try this in percolator?

**Comment by lazear** [comment]
With charge as an integer/categorical value, I get "418968 test set PSMs with q<0.01" (PXD001819). With charge one-hot encoded, I get "419171 test set PSMs with q<0.01".

I can push out another release with this change

**Comment by lazear** [comment]
@timosachsenberg new release is now available with one-hot encoding

**Comment by timosachsenberg** [comment]
Thanks. That got me one step further.

It now works in mokapot but in percolator I still get:

../THIRDPARTY/Linux/64bit/Percolator/percolator -U /home/sachsenb/results.sage.pin
Hyperparameters: selectionFdr=0.01, Cpos=0, Cneg=0, maxNiter=10
Reading tab-delimited input from datafile /home/sachsenb/results.sage.pin
Features:

Exception caught: ERROR: Reading tab file, too few features present.

thrown in https://github.com/percolator/percolator/blob/c34cafbb2f0a96b921ab86b4626980cbf25719da/src/SetHandler.cpp#L449

**Comment by timosachsenberg** [comment]
will try with 3.06 (currently on 3.05) and report back

**Comment by timosachsenberg** [comment]
I currently get a segfault for one of my files in percolator. See https://github.com/percolator/percolator/issues/352
could be a problem in 3.06 if I did not mess anything up...

**Comment by lazear** [comment]
Are there any unresolved issues left? Seems like the previous issue was due to running without decoys, and isn't an issue with Sage's output

**Comment by timosachsenberg** [comment]
I will circle back soon... got side tracked. I think it is solved...
